packages/react/src/components/Tree/Tree.tsx

```diff
@@ -0,0 +1,334 @@
+      e: React.SyntheticEvent,
+      treeItemProps: TreeItemProps,
+      predefinedProps: TreeItemProps,
+    ) => {
+      const { activeItemIds } = this.state
+      const { id, siblings } = treeItemProps
+      const { exclusive } = this.props
+
+      if (!Tree.isSubtree(treeItemProps)) {
+        return
+      }
+
+      const indexOfActiveItem = activeItemIds.indexOf(id)
+
+      if (indexOfActiveItem > -1) {
+        activeItemIds.splice(indexOfActiveItem, 1)
+      } else {
+        if (exclusive) {
+          siblings.some(sibling => {
+            const activeSiblingIndex = activeItemIds.indexOf(sibling['id'])
+            if (activeSiblingIndex > -1) {
+              activeItemIds.splice(activeSiblingIndex, 1)
+              return true
+            }
+            return false
+          })
+        }
+
+        activeItemIds.push(id)
+      }
+
+      this.setState({
+        activeItemIds,
+      })
+
+      _.invoke(predefinedProps, 'onTitleClick', e, treeItemProps)
+    },
+    onFocusParent: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { parent } = treeItemProps
+
+      if (!parent) {
+        return
+      }
+
+      const { itemsForRender } = this.state
+      const elementToBeFocused = itemsForRender[parent['id']].elementRef
+
+      if (!elementToBeFocused) {
+        return
+      }
+
+      elementToBeFocused.current.focus()
+      _.invoke(predefinedProps, 'onFocusParent', e, treeItemProps)
+    },
+    onFocusFirstChild: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { id } = treeItemProps
+
+      const { itemsForRender } = this.state
+      const currentElement = itemsForRender[id].elementRef
+
+      if (!currentElement && currentElement.current) {
+        return
+      }
+
+      const elementToBeFocused = getNextElement(this.treeRef.current, currentElement.current)
+
+      if (!elementToBeFocused) {
+        return
+      }
+
+      elementToBeFocused.focus()
+      _.invoke(predefinedProps, 'onFocusFirstChild', e, treeItemProps)
+    },
+    onSiblingsExpand: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      if (this.props.exclusive) {
+        return
+      }
+
+      const { id, siblings } = treeItemProps
+      const { activeItemIds } = this.state
+
+      siblings.forEach(sibling => {
+        if (Tree.isSubtree(sibling) && activeItemIds.indexOf(sibling['id']) < 0) {
+          activeItemIds.push(sibling['id'])
+        }
+      })
+
+      if (Tree.isSubtree(treeItemProps) && activeItemIds.indexOf(id) < 0) {
+        activeItemIds.push(id)
+      }
+
+      this.setState({
+        activeItemIds,
+      })
+
+      _.invoke(predefinedProps, 'onSiblingsExpand', e, treeItemProps)
+    },
+  })
+
+  renderContent() {
+    const { activeItemIds, itemsForRender } = this.state
+    const { items, renderItemTitle } = this.props
+
+    if (!items) return null
+
+    const renderItems = (items: ShorthandCollection<TreeItemProps>): any[] => {
```

Comment:
```tsx
    const renderItems = (items: ShorthandCollection<TreeItemProps>): any[] /* !!! */ => {
```

Why this is any? Let's type it properly

Reply:
However, I am not fully understand why we need to have there a callback.

Why we can do simply return?

```tsx
  renderContent() {
    const { activeItemIds, itemsForRender } = this.state
    const { items, renderItemTitle } = this.props

    if (!items) return null
    return items.reduce((renderedItems: any[], item: ShorthandValue<TreeItemProps>) => {
    //
```

Reply:
in List and Menu our `renderItems` are typed as `any[]`

Reply:
```tsx
    const renderItems = (items: ShorthandCollection<TreeItemProps>): React.ReactElement[] => {
      return items.reduce((renderedItems: React.ReactElement[], item: ShorthandValue<TreeItemProps>) => {
```

We checked and this works properly, it hightlights another issue typings under factories will create an issue for it.

Reply:
https://github.com/stardust-ui/react/issues/1865